File: python_server/auth/dependencies.py
from typing import Optional

from fastapi import Depends, Header, Request

from .current_user import CurrentUser
from .exceptions import FORBIDDEN, UNAUTHORIZED
from .jwt import JWTVerificationError, verify_jwt
from .service_key import ServiceKeyError, verify_service_key


def _parse_bearer(header: str) -> tuple[str, str]:
    parts = header.split(" ", 1)
    if len(parts) != 2:
        return "", ""
    return parts[0].strip(), parts[1].strip()


async def require_jwt_or_service_key(
    request: Request,
    authorization: Optional[str] = Header(None),
    x_service_key: Optional[str] = Header(None),
) -> Optional[CurrentUser]:
    """Return CurrentUser for a valid JWT, None for a valid S2S key, or raise 401."""
    settings = getattr(request.app.state, "settings", None)
    if not settings:
        raise UNAUTHORIZED

    if x_service_key is not None:
        try:
            verify_service_key(x_service_key, settings.secret_key)
            return None
        except ServiceKeyError:
            raise UNAUTHORIZED

    if not authorization:
        raise UNAUTHORIZED

    scheme, token = _parse_bearer(authorization)
    if scheme.lower() != "bearer" or not token:
        raise UNAUTHORIZED

    try:
        return verify_jwt(token, settings.jwt_secret)
    except JWTVerificationError:
        raise UNAUTHORIZED


async def require_jwt(
    request: Request,
    authorization: Optional[str] = Header(None),
) -> CurrentUser:
    """Return CurrentUser for a valid JWT or raise 401."""
    settings = getattr(request.app.state, "settings", None)
    if not settings:
        raise UNAUTHORIZED

    if not authorization:
        raise UNAUTHORIZED

    scheme, token = _parse_bearer(authorization)
    if scheme.lower() != "bearer" or not token:
        raise UNAUTHORIZED

    try:
        return verify_jwt(token, settings.jwt_secret)
    except JWTVerificationError:
        raise UNAUTHORIZED


def require_ability(ability: str):
    """Factory that returns a FastAPI dependency requiring a verified JWT with the ability."""

    async def _check(user: CurrentUser = Depends(require_jwt)) -> CurrentUser:
        if user.has_ability(ability):
            return user
        raise FORBIDDEN

    return _check

File: server/auth/dependencies.ts
import { Request, Response, NextFunction } from 'express'
import { verifyJwt } from './jwt'
import { verifyServiceKey } from './serviceKey'
import { CurrentUser } from './currentUser'

declare global {
  namespace Express {
    interface Request {
      currentUser?: CurrentUser
      serviceAuth?: boolean
    }
  }
}

function parseBearer(header: string): [string, string] {
  const idx = header.indexOf(' ')
  if (idx === -1) {
    return [header.trim(), '']
  }
  return [header.slice(0, idx).trim(), header.slice(idx + 1).trim()]
}

export function requireJwtOrServiceKey(req: Request, res: Response, next: NextFunction): void {
  const authHeader = req.headers.authorization
  const serviceKey = req.headers['x-service-key'] as string | undefined
  const secret = process.env.SECRET_KEY || ''
  const jwtSecret = process.env.JWT_SECRET

  if (serviceKey) {
    try {
      verifyServiceKey(serviceKey, secret)
      req.serviceAuth = true
      next()
      return
    } catch {
      res.status(401).json({ success: false, error: 'Authentication required' })
      return
    }
  }

  if (!authHeader) {
    res.status(401).json({ success: false, error: 'Authentication required' })
    return
  }

  const [scheme, token] = parseBearer(authHeader)
  if (scheme.toLowerCase() !== 'bearer' || !token) {
    res.status(401).json({ success: false, error: 'Authentication required' })
    return
  }

  try {
    req.currentUser = verifyJwt(token, jwtSecret)
    next()
    return
  } catch {
    res.status(401).json({ success: false, error: 'Authentication required' })
    return
  }
}

export function requireAbility(ability: string) {
  return (req: Request, res: Response, next: NextFunction): void => {
    if (req.serviceAuth) {
      res.status(403).json({ success: false, error: 'Forbidden' })
      return
    }

    const user = req.currentUser
    if (!user) {
      res.status(401).json({ success: false, error: 'Authentication required' })
      return
    }

    if (user.hasAbility(ability)) {
      next()
      return
    }

    res.status(403).json({ success: false, error: 'Forbidden' })
    return
  }
}
